## Movies Case Study

To find some interesting insights into a few movies released between 1916 and 2016, using Python.

In [127]:
# Importing the numpy and pandas packages

import numpy as np
import pandas as pd

### Reading and Inspection

**Import and read**


In [128]:
# Writing code for importing the csv file here
movies = pd.read_csv("movies.csv")
movies[movies["language"]=="English"].count()

color                        3669
director_name                3671
num_critic_for_reviews       3670
duration                     3670
director_facebook_likes      3671
actor_3_facebook_likes       3665
actor_2_name                 3670
actor_1_facebook_likes       3671
gross                        3671
genres                       3671
actor_1_name                 3671
movie_title                  3671
num_voted_users              3671
cast_total_facebook_likes    3671
actor_3_name                 3665
facenumber_in_poster         3665
plot_keywords                3644
movie_imdb_link              3671
num_user_for_reviews         3671
language                     3671
country                      3671
content_rating               3643
budget                       3671
title_year                   3671
actor_2_facebook_likes       3670
imdb_score                   3671
aspect_ratio                 3607
movie_facebook_likes         3671
dtype: int64

**Inspect the dataframe**



In [129]:
# Writing code for inspection here
movies.info()
movies.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

(3853, 28)

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [130]:
a = movies.isnull().sum()
a[a>0]

color                      2
num_critic_for_reviews     1
duration                   1
actor_3_facebook_likes     6
actor_2_name               1
actor_3_name               6
facenumber_in_poster       6
plot_keywords             30
language                   3
content_rating            48
actor_2_facebook_likes     1
aspect_ratio              72
dtype: int64

### Data Cleaning

**Drop unecessary columns**

For this case study,I will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So dropping the following columns:
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [131]:
movies.drop(["plot_keywords"],axis = 1, inplace= True)

In [132]:
# Writing code for dropping the columns here.
movies.drop(["color","director_facebook_likes","actor_1_facebook_likes","actor_2_facebook_likes","actor_3_facebook_likes","actor_2_name","cast_total_facebook_likes","actor_3_name","duration","facenumber_in_poster","content_rating","country","movie_imdb_link","aspect_ratio"],axis=1)

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  
0                    33000  
1                        0  
2                    85000  
3                   164000  
4                    24000  
...                    ...  
3848                 19000  
3849                    74  
3850                     0  
3851                   413  
3852                   456  

[3853 rows x 13 columns]

**Inspect Null values**

Finding out number of null values in each column of the dataframe 'movies'. 

In [133]:
a = movies.isnull().sum()
a[a>0]

color                      2
num_critic_for_reviews     1
duration                   1
actor_3_facebook_likes     6
actor_2_name               1
actor_3_name               6
facenumber_in_poster       6
language                   3
content_rating            48
actor_2_facebook_likes     1
aspect_ratio              72
dtype: int64

**Filling NaN values**

The `language` column has some NaN values.It is safe to replace all the missing values with `'English'`.

In [134]:
# Writing your code for filling the NaN values in the 'language' column here
movies.loc[pd.isnull(movies['language']), ['language']] = 'English'

In [135]:
(movies.language == 'English').sum()

3674

 The count of movies made in English language after replacing the NaN values with English is 3674.

### Data Analysis

**Changing the unit of columns**

Converting the unit of the `budget` and `gross` columns from `$` to `million $`.

In [136]:
# Writing code for unit conversion here
movies.budget = movies.budget/1000000
movies.gross= movies.gross/1000000
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes       gross                           genres  ...  \
0                  1000.0  760.505847  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309.404152         Action|Adventure|Fantasy  ...   
2                 11000.0  200.074175        Action|Adventure|Thriller  ...   
3                 27000.0  448.130642                  Action|Thriller  ...   
4                   640.0   73.058679          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating budget  title_year  \
0               3054.0  English      USA           PG-13  237.0      2009.0   
1               1238.0  English      USA           PG-13  300.0      2007.0   
2                994.0  English       UK           PG-13  245.0      2015.0   
3               2701.0  English      USA           PG-13  250.0      2012.0   
4                738.0  English      USA           PG-13  263.7      2012.0   

  actor_2_facebook_likes  imdb_score aspect_ratio movie_facebook_likes  
0                  936.0         7.9         1.78                33000  
1                 5000.0         7.1         2.35                    0  
2                  393.0         6.8         2.35                85000  
3                23000.0         8.5         2.35               164000  
4                  632.0         6.6         2.35                24000  

[5 rows x 27 columns]

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [137]:
# Writing code for creating the profit column here
movies["profit"] = movies["gross"]-movies["budget"]
movies[["profit"]]

profit
0     523.505847
1       9.404152
2     -44.925825
3     198.130642
4    -190.641321
...          ...
3848    0.417760
3849    0.063071
3850    2.033920
3851   -0.004416
3852    0.084122

[3853 rows x 1 columns]

In [138]:
# Code for sorting the dataframe 
movies_sort = movies.sort_values(by ="profit",ascending = False)

In [139]:
# Writing code to get the top 10 profiting movies
top10 = movies_sort.head(10) 
top10

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
28    Color    Colin Trevorrow                   644.0     124.0   
25    Color      James Cameron                   315.0     194.0   
2704  Color       George Lucas                   282.0     125.0   
2748  Color   Steven Spielberg                   215.0     120.0   
16    Color        Joss Whedon                   703.0     173.0   
482   Color       Roger Allers                   186.0      73.0   
230   Color       George Lucas                   320.0     136.0   
64    Color  Christopher Nolan                   645.0     152.0   
419   Color          Gary Ross                   673.0     142.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                         0.0                   855.0   Joel David Moore   
28                      365.0                  1000.0         Judy Greer   
25                        0.0                   794.0       Kate Winslet   
2704                      0.0                   504.0      Peter Cushing   
2748                  14000.0                   548.0        Dee Wallace   
16                        0.0                 19000.0  Robert Downey Jr.   
482                      28.0                   847.0        Nathan Lane   
230                       0.0                  1000.0        Liam Neeson   
64                    22000.0                 11000.0       Heath Ledger   
419                     378.0                   575.0    Josh Hutcherson   

      actor_1_facebook_likes       gross  \
0                     1000.0  760.505847   
28                    3000.0  652.177271   
25                   29000.0  658.672302   
2704                 11000.0  460.935665   
2748                   861.0  434.949459   
16                   26000.0  623.279547   
482                   2000.0  422.783777   
230                  20000.0  474.544677   
64                   23000.0  533.316061   
419                  34000.0  407.999255   

                                        genres  ... language country  \
0              Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
28            Action|Adventure|Sci-Fi|Thriller  ...  English     USA   
25                               Drama|Romance  ...  English     USA   
2704           Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
2748                             Family|Sci-Fi  ...  English     USA   
16                     Action|Adventure|Sci-Fi  ...  English     USA   
482   Adventure|Animation|Drama|Family|Musical  ...  English     USA   
230            Action|Adventure|Fantasy|Sci-Fi  ...  English     USA   
64                 Action|Crime|Drama|Thriller  ...  English     USA   
419            Adventure|Drama|Sci-Fi|Thriller  ...  English     USA   

      content_rating  budget title_year  actor_2_facebook_likes imdb_score  \
0              PG-13   237.0     2009.0                   936.0        7.9   
28             PG-13   150.0     2015.0                  2000.0        7.0   
25             PG-13   200.0     1997.0                 14000.0        7.7   
2704              PG    11.0     1977.0                  1000.0        8.7   
2748              PG    10.5     1982.0                   725.0        7.9   
16             PG-13   220.0     2012.0                 21000.0        8.1   
482                G    45.0     1994.0                   886.0        8.5   
230               PG   115.0     1999.0                 14000.0        6.5   
64             PG-13   185.0     2008.0                 13000.0        9.0   
419            PG-13    78.0     2012.0                 14000.0        7.3   

      aspect_ratio movie_facebook_likes      profit  
0             1.78                33000  523.505847  
28            2.00               150000  502.177271  
25            2.35                26000  458.672302  
2704          2.35                33000  449.935665  
2748         

In [140]:
top10[top10["director_name"]=="James Cameron"]

color  director_name  num_critic_for_reviews  duration  \
0   Color  James Cameron                   723.0     178.0   
25  Color  James Cameron                   315.0     194.0   

    director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                       0.0                   855.0  Joel David Moore   
25                      0.0                   794.0      Kate Winslet   

    actor_1_facebook_likes       gross                           genres  ...  \
0                   1000.0  760.505847  Action|Adventure|Fantasy|Sci-Fi  ...   
25                 29000.0  658.672302                    Drama|Romance  ...   

   language country  content_rating  budget title_year  \
0   English     USA           PG-13   237.0     2009.0   
25  English     USA           PG-13   200.0     1997.0   

    actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  \
0                    936.0        7.9          1.78                33000   
25                 14000.0        7.7          2.35                26000   

        profit  
0   523.505847  
25  458.672302  

[2 rows x 28 columns]

In [141]:
top10.iloc[[0,1,2],[11]]

movie_title
0           Avatar 
28  Jurassic World 
25         Titanic

Top 3 movies who scored the maximum profit are "Avatar","Jurassic World","Titanic".

**Finding IMDb Top 250**

Creating a new dataframe `IMDb_Top_250` and storing the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also making sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also adding a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [142]:
# Writing code for extracting the top 250 movies as per the IMDb score here. 
# Naming that dataframe as 'IMDb_Top_250'
IMDB_Top = movies.sort_values(by="imdb_score",ascending = False)
IMDB_Top_250 = IMDB_Top[IMDB_Top["num_voted_users"]>25000].head(100)
IMDB_Top_250[["movie_title"]]

movie_title
1795                       The Shawshank Redemption 
3016                                  The Godfather 
2543                         The Godfather: Part II 
64                                  The Dark Knight 
325   The Lord of the Rings: The Return of the King 
...                                              ...
1650                                   Blade Runner 
868                                          Casino 
481                                A Beautiful Mind 
26                       Captain America: Civil War 
1829                                        Warrior 

[100 rows x 1 columns]

#### <font color='black'>Checking movie count into 5 buckets based on the IMDb ratings: </font>
-  <font color='black'>7.5 to 8</font>
-  <font color='black'>8 to 8.5</font>
-  <font color='black'>8.5 to 9</font>
-  <font color='black'>9 to 9.5</font>
-  <font color='black'>9.5 to 10</font>


In [143]:
IMDB_Top_250[["imdb_score"]].value_counts()

imdb_score
8.3           23
8.5           18
8.2           16
8.4           15
8.6            8
8.7            7
8.8            5
8.9            4
9.0            2
9.2            1
9.3            1
dtype: int64

In [144]:
len(IMDB_Top_250[(IMDB_Top_250["imdb_score"]>=7.5)&(IMDB_Top_250["imdb_score"]<8)])

0

In [145]:
len(IMDB_Top_250[(IMDB_Top_250["imdb_score"]>=8)&(IMDB_Top_250["imdb_score"]<8.5)])

54

In [146]:
len(IMDB_Top_250[(IMDB_Top_250["imdb_score"]>=8.5)&(IMDB_Top_250["imdb_score"]<9.0)])

42

In [147]:
len(IMDB_Top_250[(IMDB_Top_250["imdb_score"]>=9.0)&(IMDB_Top_250["imdb_score"]<9.5)])

4

In [148]:
len(IMDB_Top_250[(IMDB_Top_250["imdb_score"]>=9.5)&(IMDB_Top_250["imdb_score"]<10)])

0

The maximum number of movies hold by "8 to 8.5" bucked from *IMDb_Top_250*

**Finding the critic-favorite and audience-favorite actors**

   1. Creating three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Using only the `actor_1_name` column for extraction.Also,making sure the use of names'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Appending the rows of all these dataframes and storing them in a new dataframe named `Combined`.
   3. Grouping the combined dataframe using the `actor_1_name` column.
   4. Finding the mean of the `num_critic_for_reviews` and `num_user_for_review` and identifying the actors which have the highest mean.

In [149]:
movies.actor_1_name

0           CCH Pounder
1           Johnny Depp
2       Christoph Waltz
3             Tom Hardy
4          Daryl Sabara
             ...       
3848      Shane Carruth
3849        Ian Gamazon
3850    Carlos Gallardo
3851        Kerry Bishé
3852        John August
Name: actor_1_name, Length: 3853, dtype: object

In [150]:
# Writing your code for creating three new dataframes here
Meryl_Streep = movies[movies["actor_1_name"]=="Meryl Streep"]
Meryl_Streep # Include all movies in which Meryl_Streep is the lead

color   director_name  num_critic_for_reviews  duration  \
392   Color    Nancy Meyers                   187.0     120.0   
1038  Color   Curtis Hanson                    42.0     111.0   
1132  Color     Nora Ephron                   252.0     123.0   
1322  Color   David Frankel                   208.0     109.0   
1390  Color  Robert Redford                   227.0      92.0   
1471  Color  Sydney Pollack                    66.0     161.0   
1514  Color   David Frankel                   234.0     100.0   
1563  Color   Carl Franklin                    64.0     127.0   
1784  Color  Stephen Daldry                   174.0     114.0   
2500  Color  Phyllida Lloyd                   331.0     105.0   
2793  Color   Robert Altman                   211.0     105.0   

      director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
392                     278.0                   963.0     Hunter Parrish   
1038                    161.0                   132.0    Glenn Morshower   
1132                      0.0                   923.0  Mary Lynn Rajskub   
1322                     64.0                   505.0      Anne Hathaway   
1390                      0.0                 10000.0         Tom Cruise   
1471                    521.0                   184.0      Michael Gough   
1514                     64.0                   329.0       Steve Carell   
1563                     73.0                   433.0       William Hurt   
1784                    335.0                   530.0    Stephen Dillane   
2500                     58.0                   583.0      Jim Broadbent   
2793                    500.0                   718.0    Virginia Madsen   

      actor_1_facebook_likes       gross                           genres  \
392                  11000.0  112.703470             Comedy|Drama|Romance   
1038                 11000.0   46.815748  Action|Adventure|Crime|Thriller   
1132                 11000.0   94.125426          Biography|Drama|Romance   
1322                 11000.0  124.732962             Comedy|Drama|Romance   
1390                 11000.0   14.998070               Drama|Thriller|War   
1471                 11000.0   87.100000          Biography|Drama|Romance   
1514                 11000.0   63.536011             Comedy|Drama|Romance   
1563                 11000.0   23.209440                            Drama   
1784                 11000.0   41.597830                    Drama|Romance   
2500                 11000.0   29.959436          Biography|Drama|History   
2793                 11000.0   20.338609               Comedy|Drama|Music   

      ... language country  content_rating  budget title_year  \
392   ...  English     USA               R    85.0     2009.0   
1038  ...  English     USA           PG-13    45.0     1994.0   
1132  ...  English     USA           PG-13    40.0     2009.0   
1322  ...  English     USA           PG-13    35.0     2006.0   
1390  ...  English     USA               R    35.0     2007.0   
1471  ...  English     USA              PG    31.0     1985.0   
1514  ...  English     USA           PG-13    30.0     2012.0   
1563  ...  English     USA               R    30.0     1998.0   
1784  ...  English     USA           PG-13    25.0     2002.0   
2500  ...  English      UK           PG-13    13.0     2011.0   
2793  ...  English     USA           PG-13    10.0     2006.0   

      actor_2_facebook_likes imdb_score  aspect_ratio movie_facebook_likes  \
392                   2000.0        6.6          1.85                    0   
1038                   894.0        6.3          2.35                    0   
1132                   935.0        7.0          1.85                13000   
1322                 11000.0        6.8          2.35                    0   
1390                 10000.0        6.2          2.35                    0   
1471                   920.0        7.2          1.85                    0   
1514                  7000.0        6.3          2.35       

In [151]:
Leo_Caprio = movies[movies["actor_1_name"]=="Leonardo DiCaprio"]# Include all movies in which Leo_Caprio is the lead

In [152]:
Brad_Pitt =  movies[movies["actor_1_name"]=="Brad Pitt"] # Include all movies in which Brad_Pitt is the lead

In [153]:
# Writing your code for combining the three dataframes here
Combined = Meryl_Streep.append(Leo_Caprio).append(Brad_Pitt)

In [154]:
# Write your code for grouping the combined dataframe here
a = Combined.groupby("actor_1_name")

In [155]:
# Write the code for finding the mean of critic reviews and audience reviews here
a.mean()

num_critic_for_reviews    duration  \
actor_1_name                                            
Brad Pitt                      245.000000  132.588235   
Leonardo DiCaprio              330.190476  148.000000   
Meryl Streep                   181.454545  115.181818   

                   director_facebook_likes  actor_3_facebook_likes  \
actor_1_name                                                         
Brad Pitt                      4633.882353              735.176471   
Leonardo DiCaprio              7453.047619             2767.476190   
Meryl Streep                    186.727273             1390.909091   

                   actor_1_facebook_likes       gross  num_voted_users  \
actor_1_name                                                             
Brad Pitt                         11000.0   66.321449    283583.823529   
Leonardo DiCaprio                 29000.0  125.741989    441958.857143   
Meryl Streep                      11000.0   59.919727     73545.545455   

                   cast_total_facebook_likes  facenumber_in_poster  \
actor_1_name                                                         
Brad Pitt                       16968.235294              1.529412   
Leonardo DiCaprio               41630.285714              1.000000   
Meryl Streep                    16426.636364              2.272727   

                   num_user_for_reviews     budget   title_year  \
actor_1_name                                                      
Brad Pitt                    742.352941  69.294118  2004.352941   
Leonardo DiCaprio            914.476190  82.452381  2005.476190   
Meryl Streep                 297.181818  34.454545  2003.545455   

                   actor_2_facebook_likes  imdb_score  aspect_ratio  \
actor_1_name                                                          
Brad Pitt                     4652.470588    7.152941      2.221176   
Leonardo DiCaprio             8719.000000    7.495238      2.254762   
Meryl Streep                  3283.636364    6.745455      2.122727   

                   movie_facebook_likes     profit  
actor_1_name                                        
Brad Pitt                  14051.764706  -2.972669  
Leonardo DiCaprio          52190.476190  43.289608  
Meryl Streep                2934.090909  25.465182

The highest rated actor among the three actors according to the user reviews is "Leonardo DiCaprio" with mean of user reviews: "915".

The highest rated actor among the three actors according to the critic reviews is "Leonardo DiCaprio" with mean of critic reviews: "331".